src/celephem/samporbit.cpp

```diff
@@ -297,135 +296,123 @@ template <typename T> Vector3d SampledOrbit<T>::computeVelocity(double jd) const
             lastSample = n;
         }
 
         if (n == 0)
         {
-            vel = Vector3d::Zero();
+            vel = Eigen::Vector3d::Zero();
         }
         else if (n < (int) samples.size())
         {
-            if (interpolation == TrajectoryInterpolationLinear)
+            switch (interpolation)
             {
-                Sample<T> s0 = samples[n - 1];
-                Sample<T> s1 = samples[n];
-
-                double dt = (s1.t - s0.t);
-                return (Vector3d(s1.x, s1.y, s1.z) - Vector3d(s0.x, s0.y, s0.z)) * (1.0 / dt);
-            }
-            if (interpolation == TrajectoryInterpolationCubic)
-            {
-                Sample<T> s0, s1, s2, s3;
-                if (n > 1)
-                    s0 = samples[n - 2];
-                else
-                    s0 = samples[n - 1];
-                s1 = samples[n - 1];
-                s2 = samples[n];
-                if (n < (int) samples.size() - 1)
-                    s3 = samples[n + 1];
-                else
-                    s3 = samples[n];
-
-                double h = s2.t - s1.t;
-                double ih = 1.0 / h;
-                double t = (jd - s1.t) * ih;
-                Vector3d p0(s1.x, s1.y, s1.z);
-                Vector3d p1(s2.x, s2.y, s2.z);
-
-                Vector3d v10((double) s1.x - (double) s0.x,
-                             (double) s1.y - (double) s0.y,
-                             (double) s1.z - (double) s0.z);
-                Vector3d v21((double) s2.x - (double) s1.x,
-                             (double) s2.y - (double) s1.y,
-                             (double) s2.z - (double) s1.z);
-                Vector3d v32((double) s3.x - (double) s2.x,
-                             (double) s3.y - (double) s2.y,
-                             (double) s3.z - (double) s2.z);
-
-                // Estimate velocities by averaging the differences at adjacent spans
-                // (except at the end spans, where we just use a single velocity.)
-                Vector3d v0;
-                if (n > 1)
-                {
-                    v0 = v10 * (0.5 / (s1.t - s0.t)) + v21 * (0.5 * ih);
-                    v0 *= h;
-                }
-                else
-                {
-                    v0 = v21;
-                }
-
-                Vector3d v1;
-                if (n < (int) samples.size() - 1)
-                {
-                    v1 = v21 * (0.5 * ih) + v32 * (0.5 / (s3.t - s2.t));
-                    v1 *= h;
-                }
-                else
-                {
-                    v1 = v21;
-                }
-
-                vel = cubicInterpolateVelocity(p0, v0, p1, v1, t);
-                vel *= 1.0 / h;
-            }
-            else
-            {
-                // Unknown interpolation type
-                vel = Vector3d::Zero();
+            case TrajectoryInterpolation::Linear:
+                vel = computeVelocityLinear(jd, n);
```

Comment:
In line 322 of the original, this value was immediately returned, skipping the z/-y swap. The old way looks incorrect to me, but need to check to make sure.

Reply:
Ok, the original code was incorrect (velocity vector reference mark displayed at 90° to the actual velocity for linear interpolated xyz files)